third_party/move/move-compiler-v2/src/pipeline/control_flow_graph_simplifier.rs

```diff
@@ -0,0 +1,599 @@
+        Self { data }
+    }
+
+    /// Does the first four transformations described in the module doc
+    fn transform(&mut self) {
+        // eliminates the trick case where a block may contain repeated successor/predecessor blocks
+        self.eliminate_branch_to_same_target();
+        let cfg1 = ControlFlowGraphCodeGenerator::new(std::mem::take(&mut self.data.code));
+        // may introduce new empty blocks, so perform before we remove empty blocks
+        let cfg2 = RedundantJumpRemover::transform(cfg1);
+        let cfg3 = EmptyBlockRemover::transform(cfg2);
+        self.data.code = cfg3.gen_code(true);
+        // the above transformation may introduce branches to same target
+        // for instance, if the original CFG is BB1 -> BB2, BB1 -> BB3 -> BB2, where BB3 is an empty block
+        self.eliminate_branch_to_same_target();
+        self.data.annotations.clear()
+    }
+
+    /// Transforms `if _ goto L else goto L` to `goto L`, which is the transformation 4 in the module doc
+    fn eliminate_branch_to_same_target(&mut self) {
+        let code = std::mem::take(&mut self.data.code);
+        self.data.code = code
+            .into_iter()
+            .map(|bytecode| match bytecode {
+                Bytecode::Branch(attr_id, l0, l1, _) if l0 == l1 => Bytecode::Jump(attr_id, l0),
+                _ => bytecode,
+            })
+            .collect();
+    }
+}
+
+/// Control flow graph that owns the code for transformation
+///
+/// To do a control flow graph based transformation:
+/// 1. construct a `ControlFlowGraphCodeGenerator` from the code,
+/// 2. perform the transformation on the `ControlFlowGraphCodeGenerator`
+/// 3. generate the code back from the `ControlFlowGraphCodeGenerator`
+#[derive(Debug)]
+struct ControlFlowGraphCodeGenerator {
+    /// The control flow graph.
+    /// `BlockContent` is invalidated during transformations
+    entry_block: BlockId,
+    exit_block: BlockId,
+    /// The code of each basic block under transformation.
+    /// Entry/exit block is mapped to an empty vector.
+    code_blocks: BTreeMap<BlockId, Vec<Bytecode>>,
+    /// Maps a basic block to its successors.
+    successors: BTreeMap<BlockId, Vec<BlockId>>,
+    /// Maps a basic block to its predecessors.
+    predecessors: BTreeMap<BlockId, Vec<BlockId>>,
+}
+
+impl ControlFlowGraphCodeGenerator {
+    pub fn new(mut code: Vec<Bytecode>) -> Self {
+        let cfg = StacklessControlFlowGraph::new_forward(&code);
+        let mut code_blocks = BTreeMap::new();
+        // traverse the blocks in decreasing order of its start code index
+        for (block_id, lower, upper) in cfg
+            .blocks()
+            .into_iter()
+            .filter_map(|block_id| match cfg.content(block_id) {
+                BlockContent::Basic { lower, upper } => Some((block_id, *lower, *upper)),
+                BlockContent::Dummy => None,
+            })
+            .sorted_by_key(|(_id, _lower, upper)| *upper)
```

Comment:
Can you use `- *upper` here and avoid the `.rev()` below?

Reply:
We can't; `upper` is unsigned.